File: Algorithms/0746_Min_Cost_Climbing_Stairs/Java/MinCostClimbingStairsSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
    public int minCostClimbingStairs(int[] cost) {
        if (cost.length <= 2) return Arrays.stream(cost).min().getAsInt();
        int temp1 = 0, temp2 = 0, temp3 = 0;
        for (int i = 2; i <= cost.length; i++) {
            temp3 = Math.min(temp1 + cost[i - 2], temp2 + cost[i - 1]);
            temp1 = temp2;
            temp2 = temp3;
        }
        return temp3;
    }
}

File: Algorithms/0746_Min_Cost_Climbing_Stairs/Python/Min_Cost_Climbing_Stairs_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:

    def minCostClimbingStairs(self, cost):
        length = len(cost)
        if length <= 2: return min(cost)
        temp_1, temp_2, temp_3 = 0, 0, 0
        for i in range(2, length + 1):
            temp_3 = min(temp_2 + cost[i - 1], temp_1 + cost[i - 2])
            temp_1, temp_2 = temp_2, temp_3

        return temp_3




